### PR TITLE
== should use virtual calls

### LINKED ISSUE #260: Inconsistency between .equals() and ==
The following code

```scala
object Main {
  def getStr(): String = {
    val bytes = Array ('h'.toByte, 'o'.toByte, 'l'.toByte, 'a'.toByte)
    
    new String (bytes)
  }
  
  def main (args : Array [String]) = {
    
    val sz = getStr()
      
    val bool = "hola" == sz
    if (bool)
      fprintf (stdout, c"==\n")
    else
      fprintf (stdout, c"!=\n")

    val bool2 = "hola".equals (sz)
    if (bool2)
      fprintf (stdout, c"equals is true\n")
    else
      fprintf (stdout, c"equals is false\n")
    
    ()
  }
```
prints
```
!=
equals is true
```

and should print

```
==
equals is true
```

### PR DESCRIPTION
**Pull request by ragnr**
Before this, == is converted to call Object.equals directly. As a
result, comparing any two different objects (even if they are equal per
.equals), will always result in false.

Fixes #260

### CODE REVIEW

**Comment by ragnr** [comment]
CI failed because I did not sign the CLA. I have signed the CLA now.

File: tests/run/obj-equals/src/main/scala/ObjEquals.scala [hunk 1]
```diff
@@ -0,0 +1,11 @@
+object ObjEquals {
+
+  case class O(m: Int)
+
+  def main(args: Array[String]): Unit = {
+    val a = new O(5)
+    val b = new O(5)
+
+    assert((a==b) == true)
```
Comment (densh) [inline tests/run/obj-equals/src/main/scala/ObjEquals.scala]: No need for `== true`. That's automatically implied by `assert`.

File: nscplugin/src/main/scala/scala/scalanative/nscplugin/NirCodeGen.scala [hunk 2]
```diff
@@ -1170,7 +1170,7 @@ abstract class NirCodeGen
         right withOp Op.Comp(comp, Rt.Object, left.value, right.value)
       } else {
         val equals = genMethodCall(NObjectEqualsMethod,
-                                   statically = true,
+                                   statically = false,
```
Comment (densh) [inline nscplugin/src/main/scala/scala/scalanative/nscplugin/NirCodeGen.scala]: Nice catch, @ragnr !

**Comment by densh** [comment]
Hi @ragnr , please port your tests to the [new unit testing infrastructure](https://github.com/scala-native/scala-native/pull/263).

**Comment by ragnr** [comment]
@densh, Will do.

After fixing == to call .equals properly, .equals is failing because null.isInstanceOf[Integer] is failing. The failure is because isInstanceOf is trying to dereference null (zero pointer). I am looking into fixing this now.

**Comment by densh** [comment]
All operations on `null` are undefined behaviour at the moment. `isInstanceOf` and `asInstanceOf` are special in this regard though as the methods that don't throw but have special semantics for nulls.

**Comment by densh** [comment]
I've just opened an issue for that https://github.com/scala-native/scala-native/issues/283

**Comment by densh** [comment]
Status here: we're merging #336 soon to fix #283 in #324 that will allow us to finally merge this pr.

**Comment by densh** [comment]
Superseded by #343

**Pull request closed without merging**